x-pack/plugins/security_solution/server/lib/detection_engine/rules/get_prepackaged_rules.ts

```diff
@@ -52,7 +53,43 @@ export const validateAllPrepackagedRules = (
   });
 };
 
+/**
+ * Retrieve and validate rules that were installed from Fleet as saved objects.
+ */
+export const getFleetInstalledRules = async (
+  client: RuleAssetSavedObjectsClient
+): Promise<AddPrepackagedRulesSchemaDecoded[]> => {
+  const fleetResponse = await client.all();
+  const fleetRules = fleetResponse.map(
+    // @ts-expect-error data is too loosely typed
+    (so) => so.attributes as AddPrepackagedRulesSchema
+  );
+  return validateAllPrepackagedRules(fleetRules);
+};
+
 export const getPrepackagedRules = (
   // @ts-expect-error mock data is too loosely typed
   rules: AddPrepackagedRulesSchema[] = rawRules
-): AddPrepackagedRulesSchemaDecoded[] => validateAllPrepackagedRules(rules);
+): AddPrepackagedRulesSchemaDecoded[] => {
+  return validateAllPrepackagedRules(rules);
+};
+
+export const getLatestPrepackagedRules = async (
+  client: RuleAssetSavedObjectsClient
+): Promise<AddPrepackagedRulesSchemaDecoded[]> => {
+  // build a map of the most version of each rule
+  const prepackaged = getPrepackagedRules();
+  const ruleMap = new Map(prepackaged.map((r) => [r.rule_id, r]));
+
+  // check the rules installed via fleet and create/update if the version is newer
+  const fleetRules = await getFleetInstalledRules(client);
+  const fleetUpdates = fleetRules.filter(
+    // @ts-expect-error data ruleMap.get() is safe because of ruleMap.has() check
```

Comment:
fwiw, `@ts-expect-error` count outside of test files is increasing per release when it should be be decreasing or staying steady or rising only slightly.

Maintainers, operators, and general outsiders from security_solutions who are cutting across our solution pepper them in and then later I think that creates the situation where it feels like the norm to whip them out and use them as if it is a normal typescript thing when in reality they're considered a code smell and something that should be fixed soon if possible.

Granted in a few situations where typescript _is very wrong_ or we have to do an import it is something we cannot avoid. However, we should avoid them altogether even if typescript is slightly wrong but we can work around typescript.

Can we can this code for example to be this:

```ts
  const fleetUpdates = fleetRules.filter((r) => {
    const ruleId = ruleMap.get(r.rule_id);
    return ruleId == null || ruleId.version < r.version;
  });
```

Then that removes the `ts-expect-error`.

For example, a maintainer can now change that code to have accidental character addition like this below and if it passes the unit tests it will end up shipping broken.
```ts
  const fleetUpdates = fleetRules.filter(
    // @ts-expect-error data ruleMap.get() is safe because of ruleMap.has() check
    (r) => !ruleMap_extra_chars.has(r.rule_id) || ruleMap.get(r.rule_id).version < r.version
  );
```

Even in tests I try to avoid them if I can. My scale of good to bad goes like this:
* When possible use regular typescript/javascript, including generics, templates, etc...
* Ok, I have to work a bit around typescript with a change in code, but I still get type safety.
* Ok, if I have to use an "as cast" like "as <something>" I will, but I tried to avoid it.
* Ok, I have to add a bang, "!" to turn something off. I really tried not to and don't want to but it's needed. I did try to use `?` or another technique but failed.
* Ok, if I have to use an "as unknown as case", things are getting worse, but still not too bad, and the alternatives are worse than just doing it in the code.
* Finally, I really have no choice here, I must use a ts-expect-error and hope that it goes away very soon. I understand it's easy to break things :-( in this area of code by accident.

Reply:
thanks for the explanation, that's good to understand more of your thought process and approach!
i am definitely not a fan of `ts-expect-error` style suppression for a whole line either. especially since a future error could be introduced and it's already suppressed. that's not good.

in this case, it seems the typesystem was unable to infer how `has` and `get` relate to each other, so even though what i did looks safe it guessed wrong.

storing `.get()` in a variable seems like just the right solution. you can explicitly check that against `null`, and then the other check it knows is non-null. i'm trying that change out locally now and running through the type checking

Reply:
Is `== null` or `=== undefined` more idiomatic in this case?

Reply:
updated in 3df98addc33f7f6b38e2aa6132426980584a4fb5